### PR TITLE
Support data citation improvement

### PR DESCRIPTION
**Pull request by LeonardoGonzales**

### CODE REVIEW

File: core-domain/src/main/java/org/uniprot/core/publication/impl/AbstractMappedReference.java [hunk 1]
```diff
@@ -24,7 +25,7 @@ public AbstractMappedReference(
             UniProtKBAccession uniProtKBAccession,
             Set<String> sourceCategories) {
         this.source = source;
-        this.pubMedId = pubMedId;
+        this.citationId = pubMedId;
         this.uniProtKBAccession = uniProtKBAccession;
```
Comment (ahmadshadab) [inline core-domain/src/main/java/org/uniprot/core/publication/impl/AbstractMappedReference.java:29]: Minor change param of constructor name from pubMedId to citationId

File: core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java [hunk 2]
```diff
@@ -35,6 +35,80 @@
...
+        TestableCitation citation =
+                new TestableCitationBuilder()
+                        .title(title)
+                        .publicationDate(publicationDate)
+                        .authoringGroupsSet(authoringGroup)
+                        .authorsSet(authors)
+                        .build();
+        assertThat(citation.getId(), is("CI-7TL5R4DGOC4O8"));
+    }
+
```
Comment (ahmadshadab) [inline core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java:111]: I think, we should add a test when citationType is null

File: core-parser/src/main/java/org/uniprot/core/parser/tsv/literature/LiteratureEntryValueMapper.java [hunk 3]
```diff
@@ -19,43 +19,57 @@
...
+        map.put("publication_date", getPublication(literature));
+        if (literature instanceof JournalArticle) {
+            JournalArticle journalArticle = (JournalArticle) literature;
+            map.put("journal", getJournal(journalArticle));
+            map.put("first_page", getFirstPage(journalArticle));
+            map.put("last_page", getLastPage(journalArticle));
+            map.put("volume", getVolume(journalArticle));
+            map.put("reference", getReference(journalArticle));
+        } else {
+            map.put("journal", "");
```
Comment (ahmadshadab) [inline core-parser/src/main/java/org/uniprot/core/parser/tsv/literature/LiteratureEntryValueMapper.java:39]: Insert few line breaks

**Review by eddturner — APPROVED** [review]

File: controlled-vocabulary/src/main/java/org/uniprot/cv/FileParseException.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    private static final long serialVersionUID = -557508108080374204L;
+
+    public FileParseException(String message) {
+        super(message);
+    }
+
+    public FileParseException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply controlled-vocabulary/src/main/java/org/uniprot/cv/FileParseException.java:18]: I did not find... Do you suggest to use any other?

File: core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java (hunk 2 above)
Comment (LeonardoGonzales) [reply core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java:111]: I think Citation Type can not be null. See BookImpl constructor for example. We have the citationType in the constructor and it happens with all concrete class under citation interface. So, this is an use case that is impossible to replicate. Does it make sense to you?

File: core-domain/src/main/java/org/uniprot/core/citation/impl/AbstractCitationImpl.java [hunk 5]
```diff
@@ -95,6 +104,55 @@ public boolean hasPublicationDate() {
...
+                        getDatabaseId(CitationDatabase.AGRICOLA)
+                                .orElseGet(
+                                        () ->
+                                                getDatabaseId(CitationDatabase.DOI)
+                                                        .map(this::generateHash)
+                                                        .orElseGet(
+                                                                () ->
+                                                                        generateHash(
+                                                                                getHashInput()))));
+    }
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply core-domain/src/main/java/org/uniprot/core/citation/impl/AbstractCitationImpl.java:141]: Can you please write your proposed suggestion? I think I did not understand how to do it

File: core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java (hunk 2 above)
Comment (ahmadshadab) [reply core-domain/src/test/java/org/uniprot/core/citation/impl/AbstractCitationBuilderTest.java:111]: Yes, it will work with existing impl but if we need new impl and pass null, it will fail. 
We don't have validation, that is why i thought to have a defensive code

File: controlled-vocabulary/src/main/java/org/uniprot/cv/FileParseException.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahmadshadab) [reply controlled-vocabulary/src/main/java/org/uniprot/cv/FileParseException.java:18]: `protected String generateId() {
        return getDatabaseId(CitationDatabase.PUBMED)
                .orElse(getDatabaseId(CitationDatabase.AGRICOLA)
                        .orElseGet(getCitationId()));
    }
    private Supplier<String> getCitationId() {
        return () ->
                getDatabaseId(CitationDatabase.DOI)
                        .map(this::generateHash)
                        .orElseGet(
                                () ->
                                        generateHash(
                                                getHashInput()));
    }`

File: core-domain/src/main/java/org/uniprot/core/citation/impl/AbstractCitationImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahmadshadab) [reply core-domain/src/main/java/org/uniprot/core/citation/impl/AbstractCitationImpl.java:141]: protected String generateId() {
        return getDatabaseId(CitationDatabase.PUBMED)
                .orElse(getDatabaseId(CitationDatabase.AGRICOLA)
                        .orElseGet(getCitationId()));
    }
    private Supplier<String> getCitationId() {
        return () ->
                getDatabaseId(CitationDatabase.DOI)
                        .map(this::generateHash)
                        .orElseGet(
                                () ->
                                        generateHash(
                                                getHashInput()));
    }
